hail/src/main/scala/is/hail/annotations/RegionPool.scala

```diff
@@ -13,6 +17,151 @@ object RegionPool {
+  val chunksEncountered = mutable.Map[Long, Long]()
+  val minSpaceRequirements = .9
+  var chunksRequested = 0
+  var cacheHits = 0
+  var deallocationsToFit = 0
+  var smallChunkCacheSize = 0
+  val smallChunkCache = new Array[LongArrayBuilder](highestSmallChunkPowerOf2 + 1)
+   (0 until highestSmallChunkPowerOf2 + 1).foreach(index => {
+     smallChunkCache(index) = new LongArrayBuilder()
+   })
+
+  def getChunkSize(chunkPointer: Long): Long =  chunksEncountered(chunkPointer)
+   def freeInLongArrayBuilder(pool: RegionPool, ab: LongArrayBuilder):Unit = {
+     val pointer = ab.pop()
+     val size = chunksEncountered(pointer)
+     pool.incrementAllocatedBytes(-1 * size)
+     freer(pointer)
+     chunksEncountered -= pointer
+   }
+  def deallocateChunksToFit(pool: RegionPool, sizeToFit: Long): Unit = {
+    deallocationsToFit += 1
+    var smallChunkIndex = highestSmallChunkPowerOf2
+    while((sizeToFit + pool.getTotalAllocatedBytes) > pool.getHighestTotalUsage &&
+          smallChunkIndex >= 0 && !chunksEncountered.isEmpty) {
+      if (!bigChunkCache.isEmpty) {
+        val toFree = bigChunkCache.lastEntry()
+        freeInLongArrayBuilder(pool, toFree.getValue)
+        if (toFree.getValue.size == 0) bigChunkCache.remove(toFree.getKey)
+      }
+      else {
+        if (smallChunkCacheSize == 0) smallChunkIndex = -1
+        else {
+          val toFree = smallChunkCache(smallChunkIndex)
+          if (toFree.size != 0) {
+            freeInLongArrayBuilder(pool, toFree)
+            smallChunkCacheSize -= 1
+          }
+          if (toFree.size == 0) smallChunkIndex -= 1
+        }
+      }
+    }
+  }
+   def newChunk(pool: RegionPool, size: Long): Long = {
+     if ((size + pool.getTotalAllocatedBytes) > pool.getHighestTotalUsage) {
+       deallocateChunksToFit(pool, size)
+     }
+     val newChunkPointer = allocator(size)
+     chunksEncountered += (newChunkPointer -> size)
+     pool.incrementAllocatedBytes(size)
+     newChunkPointer
+   }
+  def freeAll(pool: RegionPool): Unit = {
+    if (!chunksEncountered.isEmpty) {
+      smallChunkCache.foreach(ab => {
+        while (ab.size > 0) {
+          freeInLongArrayBuilder(pool, ab)
+          smallChunkCacheSize -= 1
+      }})
+      //BiConsumer needed to work with scala 2.11.12
+      bigChunkCache.forEach(new BiConsumer[Long, LongArrayBuilder]() {
+        def accept(key: Long, value: LongArrayBuilder): Unit =
+          while (value.size > 0) freeInLongArrayBuilder(pool, value)
+      })
+    }
+  }
+  def getUsage(): (Int, Int, Int) = {
+    (chunksRequested, cacheHits, deallocationsToFit)
+  }
+  def indexInSmallChunkCache(size: Long): Int = {
+    var closestPower = highestSmallChunkPowerOf2
+    while((size >> closestPower) != 1) closestPower = closestPower - 1
+    if (size % (1 << closestPower) != 0) closestPower +=1
+    closestPower
+  }
+  override def getChunk(pool: RegionPool, size: Long): (Long, Long) = {
```

Comment:
Leave new lines between method names. Necessary for the whole file really, I just highlight some examples.